crates/bevy_ecs/src/component/required.rs

```diff
@@ -306,232 +110,1462 @@ impl RequiredComponentConstructor {
+    ///
+    /// # Safety
     ///
-    /// [`Bundle`]: crate::bundle::Bundle
-    pub(crate) fn remove_explicit_components(&mut self, components: &[ComponentId]) {
-        for component in components {
-            self.0.remove(component);
+    /// - `requiree` must have been registered in `self`
+    /// - all components in `required_components` must have been registered in `self`.
+    pub(crate) unsafe fn register_required_by(
+        &mut self,
+        requiree: ComponentId,
+        required_components: &RequiredComponents,
+    ) {
+        for &required in required_components.all.keys() {
+            // SAFETY: the caller guarantees that all components in `required_components` have been registered in `self`.
+            let required_by = unsafe { self.get_required_by_mut(required).debug_checked_unwrap() };
+            required_by.insert(requiree);
         }
     }
 
-    /// Merges `required_components` into this collection. This only inserts a required component
-    /// if it _did not already exist_ *or* if the required component is more specific than the existing one
-    /// (in other words, if the inheritance depth is smaller).
+    /// Registers the given component `R` and [required components] inherited from it as required by `T`.
     ///
-    /// See [`register_dynamic_with`](Self::register_dynamic_with) for details.
-    pub(crate) fn merge(&mut self, required_components: &RequiredComponents) {
-        for (
-            component_id,
-            RequiredComponent {
-                constructor,
-                inheritance_depth,
-            },
-        ) in required_components.0.iter()
-        {
-            // SAFETY: This exact registration must have been done on `required_components`, so safety is ensured by that caller.
-            unsafe {
-                self.register_dynamic_with(*component_id, *inheritance_depth, || {
-                    constructor.clone()
-                });
+    /// When `T` is added to an entity, `R` will also be added if it was not already provided.
+    /// The given `constructor` will be used for the creation of `R`.
+    ///
+    /// [required components]: Component#required-components
+    ///
+    /// # Safety
+    ///
+    /// - the given component IDs `required` and `requiree` must be valid in `self`;
+    /// - the given component ID `required` must be valid for the component type `R`.
+    ///
+    ///
+    /// # Errors
+    ///
+    /// Returns a [`RequiredComponentsError`] if either of these are true:
+    /// - the `required` component is already a *directly* required component for the `requiree`; indirect
+    ///   requirements through other components are allowed. In those cases, the more specific
+    ///   registration will be used.
+    /// - the `requiree` component is already a (possibly indirect) required component for the `required` component.
+    pub(crate) unsafe fn register_required_components<R: Component>(
```

Comment:
I haven't looked too closely at this but there's something I don't understand. 
We seem to have a typed API in register.rs that uses the function defined by the macro on the Component type. And then we have this 'untyped' api here in required.rs

How come we cannot use the `register.rs` version at all times? It seems much simpler to understand. Is it because that version can only be used for first time registration?

Reply:
I'm not sure what you mean, after this PR there's no API for registering required components in `register.rs`.

Do you perhaps mean the few lines in the middle of `register_component_unchecked`? Those indeed only work for the first time, or rather they work because we know that there's no other component requiring the current component (or we would have a cycle), but that's not an assumption that can be made here.